src/lib/reasoners/adt_rel.ml

```diff
@@ -218,334 +327,201 @@ let assume_th_elt t th_elt _ =
+  { env with domains = Domains.tighten r nd env.domains }
 
-(* For a uninterpreted semantic value [r] and [h] a constructor of the form:
-     (cons (d1 ty1) ... (dn tyn))
-   this function adds the new equation to [eqs]:
-     [t = cons x1 ... xn]
-   where x1, ..., xn are fresh names of type respectively ty1, ..., tyn
-   and [t] is the term associated with the uninterpreted semantic value [r].
+(* Update the domains of the semantic values [r1] and [r2] according to
+   the substitution `r1 |-> r2`.
 
-   If the tester [(_ cons) t] hasn't been already seen, we also add it to
-   the equations [eqs].
+   @raise Domain.Inconsistent if this substitution is inconsistent with
+          the current environment [env]. *)
+let assume_subst ~ex r1 r2 env =
+  let domains = Domains.subst ~ex r1 r2 env.domains in
+  { env with domains }
 
-   Assume that [r] is a semantic value of type [Ty.Adt]. *)
-let deduce_is_constr uf r h eqs env ex =
-  let r, ex' = try Uf.find_r uf r with Not_found -> assert false in
-  let ex = Ex.union ex ex' in
-  match Th.embed r with
-  | Adt.Alien r ->
-    begin match X.term_extract r with
-      | Some t, _ ->
-        let eqs =
-          if seen_tester r h env then eqs
-          else
-            let is_c = E.mk_tester (Hstring.view h) t in
-            if Options.get_debug_adt () then
-              Printer.print_dbg
-                ~module_name:"Adt_rel"
-                ~function_name:"deduce_is_constr"
-                "%a" E.print is_c;
-            (Literal.LTerm is_c, ex, Th_util.Other) :: eqs
-        in
-        begin
-          match E.term_view t with
-          | { E.ty = Ty.Tadt (name, params) as ty; _ } ->
-            (* Only do this deduction for finite types ??
-                 may not terminate in some cases otherwise.
-                 eg. type t = A of t
-                 goal g: forall e,e' :t. e = C(e') -> false
-                 + should not be guareded by "seen_tester"
-            *)
-            let Ty.Adt cases = Ty.type_body name params in
-            let destrs =
-              try Ty.assoc_destrs h cases with Not_found -> assert false
-            in
-            let xs = List.map (fun (_, ty) -> E.fresh_name ty) destrs in
-            let cons = E.mk_constr (Hstring.view h) xs ty in
-            let env = {env with new_terms = SE.add cons env.new_terms} in
-            let eq = E.mk_eq t cons ~iff:false in
-            if Options.get_debug_adt () then
-              Printer.print_dbg
-                ~module_name:"Adt_rel"
-                ~function_name:"deduce equal to constr"
-                "%a" E.print eq;
-            let eqs = (Literal.LTerm eq, ex, Th_util.Other) :: eqs in
-            env, eqs
-          | _ -> env, eqs
-        end
-      | _ ->
-        Printer.print_err "%a" X.print r;
-        assert false
-    end
-  | Constr _ | Select _ -> env, eqs
-
-(* Collect all the constructors of the ADT type [ty]. *)
-let values_of ty =
+let is_tightenable ty c =
```

Comment:
This looks like it checks whether `c` is an enum constructor of `ty`, is that correct? If so `is_enum_constructor` would be a more clear name, or a short comment on `is_tightenable` to explain this.

Reply:
I believe it's matter of taste. If I rename it into `is_enum_constructor`, I have to explain why we use this filter in `assume_distinct`. I suggest to rename it but I'll add a comment in `assume_distinct` too.

Reply:
I suggest this new different implementation:
```ocaml 
let assume_distinct =
  let aux ~ex r r' env =
    let d = Domains.get r env.domains in
    match Domain.as_singleton d with
    | Some (c, ex') when is_enum_cons (X.type_info r) c ->
      let ex = Ex.union ex ex' in
      let nd = Domain.remove ~ex c d in
      tighten_domain r' nd env
    | Some _ | None ->
      (* If `d` is a singleton domain but its constructor has a payload,
         we cannot tighten the domain of `r'` because they could have
         distinct payloads. *)
      env
  in fun ~ex r1 r2 env ->
    aux ~ex r1 r2 env |> aux ~ex r2 r1
```

Reply:
Actually this implementation was buggy :p  but thanks to the `distinct_same_constructor` I caught the bug. I removed the constructor from the singleton domain! 

```ocaml 
let assume_distinct =
  let aux ~ex r1 r2 env =
    let d1 = Domains.get r1 env.domains in
    let d2 = Domains.get r2 env.domains in
    match Domain.as_singleton d1 with
    | Some (c, ex') when is_enum_cons (X.type_info r1) c ->
      let ex = Ex.union ex ex' in
      let nd = Domain.remove ~ex c d2 in
      tighten_domain r2 nd env
    | Some _ | None ->
      (* If `d` is a singleton domain but its constructor has a payload,
         we cannot tighten the domain of `r'` because they could have
         distinct payloads. *)
      env
  in fun ~ex r1 r2 env ->
    aux ~ex r1 r2 env |> aux ~ex r2 r1
```